### ISSUE TITLE
`kube-aws update` fails when `workerCount` is going down to zero

### ISSUE THREAD

**Issue by andrejvanderzee** [kubernetes-incubator/kube-aws#176]
From the [documentation](https://github.com/coreos/kube-aws/blob/master/Documentation/kubernetes-on-aws-node-pool.md) on node pools I understand that node pools are created in an existing cluster. First I created this cluster:

```
clusterName: PRD
externalDNSName: xxx
createRecordSet: true
recordSetTTL: 320
hostedZoneId: "Z2N238IAN5AZMQ"
keyName: aws-vanderzee-it
region: eu-west-1
kmsKeyArn: xxx
controllerCount: 3
controllerInstanceType: t2.medium
controllerRootVolumeSize: 32
controllerRootVolumeType: gp2
workerCount: 3
subnets:
  - availabilityZone: eu-west-1a
    instanceCIDR: "10.0.0.0/24"
  - availabilityZone: eu-west-1b
    instanceCIDR: "10.0.1.0/24"
  - availabilityZone: eu-west-1c
    instanceCIDR: "10.0.2.0/24"
workerInstanceType: t2.medium
workerRootVolumeSize: 32
workerRootVolumeType: gp2
etcdCount: 3
etcdInstanceType: t2.medium
etcdRootVolumeSize: 32
etcdDataVolumeEphemeral: false
etcdDataVolumeSize: 32
vpcCIDR: "10.0.0.0/16"
serviceCIDR: "10.3.0.0/24"
podCIDR: "10.2.0.0/16"
dnsServiceIP: 10.3.0.10
mapPublicIPs: true
tlsCADurationDays: 3650
tlsCertDurationDays: 3650
kubernetesVersion: v1.4.6_coreos.0
hyperkubeImageRepo: quay.io/coreos/hyperkube
awsCliImageRepo: quay.io/coreos/awscli
awsCliTag: master
useCalico: true
containerRuntime: docker
stackTags:
  Name: "Kubernetes"
  Environment: "PRD"
```

It comes up as expected. Then, I set the workerCount to 0 and apply it to the cluster, but unfortunately this fails:

```
$ ../bin/kube-aws update --s3-uri s3://xxx/kubernetes/prd
Error: Error updating cluster: unexpected stack status: UPDATE_ROLLBACK_IN_PROGRESS
```

```$ ../bin/kube-aws version
kube-aws version v0.9.2
```

**Comment by mumoshu** [comment]
Would you mind posting recent stack events for the cluster?

**Comment by andrejvanderzee** [comment]
```
00:37:28 UTC+0100	UPDATE_ROLLBACK_COMPLETE	AWS::CloudFormation::Stack	PRD	
00:37:27 UTC+0100	UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS	AWS::CloudFormation::Stack	PRD	
00:37:26 UTC+0100	UPDATE_COMPLETE	AWS::AutoScaling::AutoScalingGroup	AutoScaleWorker	
00:37:20 UTC+0100	UPDATE_ROLLBACK_IN_PROGRESS	AWS::CloudFormation::Stack	PRD	The following resource(s) failed to update: [AutoScaleWorker].
00:37:19 UTC+0100	UPDATE_FAILED	AWS::AutoScaling::AutoScalingGroup	AutoScaleWorker	Invalid Autoscaling rolling update policy: MinInstancesInService must be greater than or equal to 0
00:37:14 UTC+0100	UPDATE_IN_PROGRESS	AWS::CloudFormation::Stack	PRD	User Initiated
00:34:53 UTC+0100	CREATE_COMPLETE	AWS::CloudFormation::Stack	PRD	
00:34:48 UTC+0100	CREATE_COMPLETE	AWS::AutoScaling::AutoScalingGroup	AutoScaleWorker	
00:33:29 UTC+0100	CREATE_IN_PROGRESS	AWS::AutoScaling::AutoScalingGroup	AutoScaleWorker	Resource creation Initiated
00:33:29 UTC+0100	CREATE_IN_PROGRESS	AWS::AutoScaling::AutoScalingGroup	AutoScaleWorker	
00:33:23 UTC+0100	CREATE_COMPLETE	AWS::AutoScaling::AutoScalingGroup	AutoScaleController	
00:27:58 UTC+0100	CREATE_COMPLETE	AWS::EC2::Instance	InstanceEtcd2	
00:27:58 UTC+0100	CREATE_COMPLETE	AWS::EC2::Instance	InstanceEtcd1	
00:27:58 UTC+0100	CREATE_COMPLETE	AWS::EC2::Instance	InstanceEtcd0	
```

**Comment by mumoshu** [comment]
Thanks for the info!

Could you try to tweak `worker.autoScalingGroup` configuration a bit?

More concretely, instead of:

```
workerCount: 0
```

how about:

```
worker:
  autoScalingGroup:
    minSize: 0
    rollingUpdateMinInstancesInService: 0
```

and then `kube-aws update`?

If that doesn't make your ASG's desired capacity down to zero, would you mind recreating your cluster with:

```
# workerCount: should be omitted and instead the below should be added before `kube-aws up`
worker:
  autoScalingGroup:
    minSize: 0
    rollingUpdateMinInstancesInService: 0
```

**Comment by mumoshu** [comment]
I've changed the issue title to reflect that update failure is happening for the main cluster, not the node pool :bow:

**Comment by andrejvanderzee** [comment]
This got rid of the error:

```
worker:
  autoScalingGroup:
    minSize: 0
    rollingUpdateMinInstancesInService: 0
```

After successful completion I still have one worker running though in `eu-west-1b`.

Getting late will resume tomorrow.

**Comment by mumoshu** [comment]
@andrejvanderzee Thanks for your support 👍 Then, recreating the main cluster like I'd explained in my prev comment would work.

It turns out we need to fix our doc at least, and revisit the process how we'd launch the first node pool, anyways. Thanks again for your contribution! I'm looking forward to your result.
Good night 🌃

**Comment by andrejvanderzee** [comment]
When I re-create the cluster like this:

```
workerCount: 0
worker:
  autoScalingGroup:
    minSize: 0
    rollingUpdateMinInstancesInService: 0
```

Then it will create a cluster with 0 workers.

When I leave out `workerCount: 0` and re-create the cluster, then it will create 1 worker.

**Comment by mumoshu** [comment]
@andrejvanderzee It seems we've been missing a strict test for that 😞
I've already updated our doc to reflect your feedback:
https://github.com/coreos/kube-aws/blob/master/Documentation/kubernetes-on-aws-node-pool.md#deploying-a-multi-az-cluster-with-cluster-autoscaler-support-with-node-pools
Thanks for the feedback!

**Comment by c-knowles** [comment]
Likely related to my changes, will see if I can add a test/fix this.

**Comment by c-knowles** [comment]
@mumoshu I had a quick look at this, the problem is really related to the defaulting going on plus golang zero-values. We have a few options to resolve:
1. Remove `workerCount` completely (you said you wanted to keep compatibility for one release)
2. Switch YAML parsers so zero-values and nil values can be distinguished
3. Default `workerCount` to zero in the code but add an uncommented `workerCount: 1` to the cluster.yaml
4. Do nothing for now until we do one of the above options later

My preference is probably 4 followed by 1 but option to doing 3 now.

**Comment by andrejvanderzee** [comment]
@mumoshu @c-knowles In the longer run, are there any plans of creating node pools in the first `up` when creating the cluster and assign etcd, worker and controllers to the node pools, i.e. enable cluster-creation in one run instead of multiple steps like is documented now?

**Comment by mumoshu** [comment]
@andrejvanderzee Not planned yet but definitely possible. Would you mind inputting us your use-case or purpose of having it?

**Comment by mumoshu** [comment]
@c-knowles Thanks for taking a look into this!

I agree to you. Anyways, I'd like to go with the option 4 for now because simply I'm not sure what is the preferred for everyone.

**Comment by andrejvanderzee** [comment]
@mumoshu My objective is to configure and run the cluster in one go instead of multiple steps. My intended cluster's etcd/controllers/workers would be spread over three AZs in private subnets and would have a public-facing ELB for the controllers. Moreover, every AZ must run an ingress controller (I understood "tainted node pools" could achieve this) also with a public-facing ELB. 

It would be great if kube-aws could facilitate in these kind of configurations in one cluster.yml and bring it up with `kube-aws up`. But surely I am not sure if this conflicts with other people's use cases :-)

**Comment by mumoshu** [comment]
@andrejvanderzee Thanks for the feedback 👍 

Sounds great but may I ask you things because I'm not yet sure the overall U/X implied with your expectation.

* Would you like to use `kube-aws update` for updating node pools after that, or for just a main cluster?
* Would you like to use `kube-aws destroy` for just destroying the main cluster or rather the whole cluster including node pools?

**Comment by andrejvanderzee** [comment]
@mumoshu At the moment, the main use-case is setup and destroy complete clusters quickly for demos and integration tests.

**Comment by mumoshu** [comment]
Hi @andrejvanderzee, fyi probably we're going to support that use-case in process of #238, specifically https://github.com/coreos/kube-aws/issues/238#issuecomment-272128944

**Comment by andrejvanderzee** [comment]
@mumoshu Great! Haven't had much time lately to get involved, but would be happy to test :-)

**Comment by whereisaaron** [comment]
@andrejvanderzee @mumoshu I am very late to this party, but I had this same problem with 0.9.3-rc.5, when setting the workers to zero, I get the error `Invalid Autoscaling rolling update policy: MinInstancesInService must be greater than or equal to 0`. I was able to work around it with the following settings. This results in a main worker pool with zero instances (I use node pools for all workers). Note that I did not need to recreate the cluster, I could use these settings with a `kube-aws update`.

```
workerCount: 0
worker:
  autoScalingGroup:
    minSize: 0
    maxSize: 0
    rollingUpdateMinInstancesInService: 1
```

I assume this is isn't a problem in 0.9.4+ but AFAIK newer versions are not forward compatible with 0.9.3 so I can't upgrade for this cluster.
